### PR TITLE
Handle setjmp/longjmp.

### PR DESCRIPTION
**Pull request by ptersilie**
We currently can't deal with setjmp/longjmp. For now ignore traces containing these calls so we don't crash.

### CODE REVIEW

**Comment by vext01** [comment]
Is the intention to do something better with setjmp/longjmp soon?

File: ykllvmwrap/src/jitmodbuilder.cc [hunk 1]
```diff
@@ -1411,6 +1411,12 @@ class JITModBuilder {
             ResumeAfter = make_tuple(CurInstrIdx, CI);
             break;
           } else {
+            StringRef S = CF->getName();
+            if (S == "_setjmp" || S == "_longjmp") {
```
Comment (vext01) [inline ykllvmwrap/src/jitmodbuilder.cc:1415]: What's with the leading `_`? Is that platform specific?

**Comment by ptersilie** [comment]
Define "soon". We could see how often we actually see this in practise, and then decide on its priority. At the moment I rather concentrate on making sure the deoptimisation is working properly but getting my tests running.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 1 above)
Comment (ptersilie) [reply ykllvmwrap/src/jitmodbuilder.cc:1415]: `_setjmp` only restores registers/stack, whereas `setjmp` restores the signal mask. I was thinking whether I should add the others too, but thought I hold out until I actually see them in the wild.

File: ykllvmwrap/src/ykllvmwrap.cc [hunk 2]
```diff
@@ -397,6 +397,11 @@ void *compileIRTrace(FN Func, char *FuncNames[], size_t BBs[], size_t TraceLen,
   std::tie(JITMod, TraceName, GlobalMappings, AOTMappingVec) =
       Func(AOTMod, FuncNames, BBs, TraceLen, FAddrKeys, FAddrVals, FAddrLen);
 
+  // If we failed to build the trace, return null.
+  if (JITMod == nullptr) {
+    return nullptr;
```
Comment (vext01) [inline ykllvmwrap/src/ykllvmwrap.cc:402]: There are going to be a few different cases where we abort trace compilation. Would it be an idea to build in (now or soon) some way of indicating *why* we couldn't build the trace?

In the test we have:
```
jit-state: invalid trace
```

I think that should eventually be something more like:
```
jit-state: abort-trace-compilation (trace uses setjmp/longjmp)
```

This could be a separate PR. It will require re-engineering of the return value (possibly using pointer tagging or a union?).

File: tests/c/setlongjmp.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Run-time:
+//   env-var: YKD_PRINT_IR=jit-pre-opt
+//   env-var: YKD_SERIALISE_COMPILATION=1
+//   env-var: YKD_PRINT_JITSTATE=1
+//   stderr:
+//     ...
+//     jit-state: stop-tracing
+//     jit-state: invalid trace
```
Comment (vext01) [inline tests/c/setlongjmp.c:8]: I think all the jit states should be hyphenated for consistency, but also see comment below about the reason for aborting.

File: ykrt/src/mt.rs [hunk 4]
```diff
@@ -392,15 +392,23 @@ impl MT {
...
+            match irtrace.compile() {
+                Ok((codeptr, di_tmpfile)) => {
+                    let ct = Box::new(CompiledTrace::new(codeptr, di_tmpfile));
+                    // FIXME: although we've now put the compiled trace into the `HotLocation`,
+                    // there's no guarantee that the `Location` for which we're compiling will ever
+                    // be executed again. In such a case, the memory has, in essence, leaked.
+                    mtx.lock().replace(ct);
+                }
+                Err(e) => {
+                    // FIXME: Properly handle failed traces, e.g. depending on the reason for the
```
Comment (vext01) [inline ykrt/src/mt.rs:404]: "failed traces" -> "failed trace compilation"?

**Comment by vext01** [comment]
> Define "soon". We could see how often we actually see this in practise, and then decide on its priority. At the moment I rather concentrate on making sure the deoptimisation is working properly but getting my tests running.

OK, so not immediately. Can you raise an issue and link to it in the code?

File: ykllvmwrap/src/ykllvmwrap.cc (hunk 2 above)
Comment (ptersilie) [reply ykllvmwrap/src/ykllvmwrap.cc:402]: It would be great to have better "errors". But funneling errors from `jitmodbuilder` via FFI back to Rust sounded a bit more involved than what I wanted to achieve with this PR.

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 1 above)
Comment (vext01) [reply ykllvmwrap/src/jitmodbuilder.cc:1415]: OK

File: ykllvmwrap/src/ykllvmwrap.cc (hunk 2 above)
Comment (vext01) [reply ykllvmwrap/src/ykllvmwrap.cc:402]: Yeah. It's a bit annoying because you would need a way to express the error inside the pointer return value.

Comment (vext01) [reply ykllvmwrap/src/ykllvmwrap.cc:402]: Raise an issue and link in FIXME?

**Comment by ptersilie** [comment]
> > Can you raise an issue and link to it in the code?

[f7cfc89](https://github.com/ykjit/yk/pull/609/commits/f7cfc895158e32a160c24eb2b27bec04b9565fde)

File: ykrt/src/mt.rs (hunk 4 above)
Comment (ptersilie) [reply ykrt/src/mt.rs]: [d7a22fe](https://github.com/ykjit/yk/pull/609/commits/d7a22fe38b37e2a8824fd8eaa9d8c89e3165e8a7)

File: tests/c/setlongjmp.c (hunk 3 above)
Comment (ptersilie) [reply tests/c/setlongjmp.c]: [44e67a7](https://github.com/ykjit/yk/pull/609/commits/44e67a77a582bd8919a16ad752e73f2451d81c07)

Comment (ptersilie) [reply tests/c/setlongjmp.c]: [875f484](https://github.com/ykjit/yk/pull/609/commits/875f484321f343f346c606a8fa577838b041f81d)

File: ykllvmwrap/src/jitmodbuilder.cc (hunk 1 above)
Comment (ltratt) [reply ykllvmwrap/src/jitmodbuilder.cc:1415]: Why not add something like `assert(S != "setjmp)` ?

File: ykllvmwrap/src/ykllvmwrap.cc (hunk 2 above)
Comment (ptersilie) [reply ykllvmwrap/src/ykllvmwrap.cc:402]: [d6951d5](https://github.com/ykjit/yk/pull/609/commits/d6951d5e61db84d131ee1f5f4c7c2222509d2c8f)

Comment (ltratt) [reply ykllvmwrap/src/ykllvmwrap.cc:402]: Good idea. I'm not really sure *what* we should do here: it probably has to be quite lightweight. Punting for now is fine.

File: ykrt/src/mt.rs [hunk 5]
```diff
@@ -392,15 +392,23 @@ impl MT {
...
-                Err(e) => todo!("{e:?}"),
+            match irtrace.compile() {
+                Ok((codeptr, di_tmpfile)) => {
+                    let ct = Box::new(CompiledTrace::new(codeptr, di_tmpfile));
+                    // FIXME: although we've now put the compiled trace into the `HotLocation`,
+                    // there's no guarantee that the `Location` for which we're compiling will ever
+                    // be executed again. In such a case, the memory has, in essence, leaked.
+                    mtx.lock().replace(ct);
+                }
+                Err(e) => {
```
Comment (ltratt) [inline ykrt/src/mt.rs:403]: Since we don't use `e` I suggest `_e` to avoid "unused variable" warnings

File: ykrt/src/mt.rs [hunk 6]
```diff
@@ -392,15 +392,23 @@ impl MT {
...
+                    // FIXME: Properly handle failed trace compilation, e.g. depending on the
+                    // reason for the failure we might want to block this location from being
+                    // traced again or only temporarily put it on hold and try again later.
+                    #[cfg(feature = "yk_jitstate_debug")]
+                    print_jit_state("trace-compilation-aborted");
+                    ()
```
Comment (ltratt) [inline ykrt/src/mt.rs:413]: This line is unnecessary.

File: ykrt/src/mt.rs (hunk 5 above)
Comment (ptersilie) [reply ykrt/src/mt.rs]: [7e47cd4](https://github.com/ykjit/yk/pull/609/commits/7e47cd4ed916b17e1a547883633ee3dfff13d13e)

File: ykrt/src/mt.rs (hunk 6 above)
Comment (ptersilie) [reply ykrt/src/mt.rs]: [7e47cd4](https://github.com/ykjit/yk/pull/609/commits/7e47cd4ed916b17e1a547883633ee3dfff13d13e)

**Comment by ptersilie** [comment]
Think that takes care of all the comments.

**Comment by vext01** [comment]
LGTM. @ltratt?

**Comment by ltratt** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed.
